qualtran/_infra/binst_graph_iterators.py

```diff
@@ -0,0 +1,100 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import heapq
+from collections import Counter
+from typing import Dict, Iterator, List, TYPE_CHECKING
+
+import attrs
+import networkx as nx
+
+if TYPE_CHECKING:
+    from qualtran import BloqInstance
+
+_INFINITY: int = int(1e18)
+
+
+def _priority(node: 'BloqInstance') -> int:
+    from qualtran._infra.gate_with_registers import total_bits
+    from qualtran._infra.quantum_graph import DanglingT
+    from qualtran.bloqs.bookkeeping import Allocate, Free
+
+    if isinstance(node, DanglingT):
+        return 0
+
+    if node.bloq_is(Allocate):
+        return _INFINITY
+
+    if node.bloq_is(Free):
+        return -_INFINITY
+
+    signature = node.bloq.signature
+    return total_bits(signature.rights()) - total_bits(signature.lefts())
+
+
+@attrs.frozen(order=True)
+class _PrioritizedItem:
+    """Helper dataclass to insert items in a heap as part of `greedy_topological_sort`."""
+
+    item: 'BloqInstance' = attrs.field(eq=_priority, order=_priority)
+    priority: int
+
+
+def greedy_topological_sort(binst_graph: nx.DiGraph) -> Iterator['BloqInstance']:
```

Comment:
fwiw networkx provides [`lexicographical_topological_sort`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.lexicographical_topological_sort.html#networkx.algorithms.dag.lexicographical_topological_sort) for this. Could pass the `priority` as the key to this.

Reply:
Great, I verified the implementation is the same so now delegating directly to `nx.lexicographical_topological_sort` for implementation.